cpp/src/arrow/compare.cc

```diff
@@ -717,6 +719,83 @@ class TypeEqualsVisitor {
   bool result_;
 };
 
+#define VISIT_SIMPLE(SCALAR_TYPE)                                 \
+  Status Visit(const SCALAR_TYPE& left) {                         \
+    const auto& right = checked_cast<const SCALAR_TYPE&>(right_); \
+    result_ = right.value == left.value;                          \
+    return Status::OK();                                          \
+  }
+
+class ScalarEqualsVisitor {
+ public:
+  explicit ScalarEqualsVisitor(const Scalar& right) : right_(right), result_(false) {}
+
+  Status Visit(const NullScalar& left) {
```

Comment:
this might give you an unused variable warning later; might want to delete `left`

```suggestion
  Status Visit(const NullScalar&) {
```